server/location/index.js

```diff
@@ -0,0 +1,175 @@
+const debug = require('debug')('kyf:matcher');
+
+// const { constituencyFromPostcode, locationInfoFromPostcode } = require ('./externals');
+const { isPc7, isPc8, isPostcodeDistrict, isFullPostcode, isPostcode, endsWithPostcode,
+pc7FromFullPostcode, pc8FromFullPostcode, districtFromFullPostcode, districtFromPostcodeDistrict, postcodeFromString,
+toStandardLatLong, toLatLong, isLeafletLatLng, latLongFromString, latLongFrom4dpLatLongString, roundToNearest } = require ('./helpers');
+const { fromPostcodesIo, fromGoogle, fromTwitter,
+  constituencyFromPostcode, locationInfoFromPostcode  } = require ('./requests');
+
+const LocationObject = ()=> ({
+  specificity : 0,
+  twitterString  : '',
+  defaultTwitterFollow : '@uklabour'
+});
+const noVerify = { verify : false };
+
+const cache = {
+  records : 0,
+
+  canonicalise : location => {
+    const latLong = toStandardLatLong(location);
+    if (latLong)
+      return(`${latLong.lng},${latLong.lat}`);                // see note in helpers.js
+    if (isPostcode(location))
+      return pc8FromFullPostcode(location) || districtFromPostcodeDistrict(location)
+    if (endsWithPostcode(location))
+      return postcodeFromString(location)
+    // and then the harder cases, eg London, Hackney, The World
+    return location
+  },
+
+  compress : x=> x,
+
+  uncompress : x=> x,
+
+  put : (location, result, options={ verify:true } ) => {
+    if (options.verify)
+      location = cache.canonicalise(location);
+    if (!cache[location]) {
+      cache[location] = cache.compress(result);
+      cache.records++;
+    }
+    // console.log(cache);
+  },
+
+  get : (location, options={ verify:true } ) => {
+    if (options.verify)
+      location = cache.canonicalise(location);
+    return cache.uncompress(cache.location);
+  }
+
+}
+
+
+// locations is one of:
+// . query object, eg {pc='AB1 9XY', latlong=<leafletJS latlong>}
+// . pcd, pc7, pc8, latLong, leafletJS latlong
+// . a prioirty ordered array of the above
+const populateLocationObject = async (locations, options={} ) => {
+  if (!Array.isArray(locations))
+    locations = [locations];
+
+  const possibles = locations
+    .map (async (location, idx) => {
+      const result = LocationObject();
+      if (!location)
+        return null
+      // NB currently does not cache failures well
+
+  // console.log('\ntrying to convert ',idx,location);
+      if (typeof location !=='object') {
+        location = cache.canonicalise(location);
+
+  if (!cache[location])
+  {} // console.log('Not cached:',location)
```

Comment:
this could be tidied up, if you're not going to log anything here. you could also use `debug()`